posthog-server/src/main/java/com/posthog/server/internal/FlagEvaluator.kt

```diff
@@ -0,0 +1,803 @@
+import java.time.temporal.ChronoUnit
+import java.util.Date
+import java.util.regex.PatternSyntaxException
+
+/**
+ * Local evaluation engine for feature flags
+ */
+internal class FlagEvaluator(
+    private val config: PostHogConfig,
+) {
+    companion object {
+        private const val LONG_SCALE = 0xFFFFFFFFFFFFFFF.toDouble()
+        private val NONE_VALUES_ALLOWED_OPERATORS = setOf("is_not")
+        private val REGEX_COMBINING_MARKS = "\\p{M}+".toRegex()
+
+        // Date formatters for parsing various date formats
+        private val DATE_FORMATTER_WITH_SPACE_TZ = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss XXX")
+        private val DATE_FORMATTER_NO_SPACE_TZ = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ssXXX")
+        private val DATE_FORMATTER_NO_TZ = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+
+        private fun casefold(input: String): String {
+            val normalized = Normalizer.normalize(input, Normalizer.Form.NFD)
+            return REGEX_COMBINING_MARKS.replace(normalized, "").uppercase().lowercase()
+        }
+    }
+
+    private data class VariantLookupEntry(
+        val key: String,
+        val valueMin: Double,
+        val valueMax: Double,
+    )
+
+    /**
+     * Hash function for consistent rollout percentages
+     * Given the same distinct_id and key, it'll always return the same float.
+     * These floats are uniformly distributed between 0 and 1.
+     */
+    private fun hash(
+        key: String,
+        distinctId: String,
+        salt: String = "",
+    ): Double {
+        val hashKey = "$key.$distinctId$salt"
+        val digest = MessageDigest.getInstance("SHA-1")
+        val hashBytes = digest.digest(hashKey.toByteArray(Charsets.UTF_8))
+
+        // Take first 15 hex characters (60 bits)
+        val hexString = hashBytes.joinToString("") { "%02x".format(it) }
+        val hashValue = hexString.substring(0, 15).toLong(16)
+
+        return hashValue / LONG_SCALE
+    }
+
+    /**
+     * Get the matching variant for a multivariate flag
+     */
+    fun getMatchingVariant(
+        flag: FlagDefinition,
+        distinctId: String,
+    ): String? {
+        val hashValue = hash(flag.key, distinctId, salt = "variant")
+        val variants = variantLookupTable(flag)
+
+        for (variant in variants) {
+            if (hashValue >= variant.valueMin && hashValue < variant.valueMax) {
+                return variant.key
+            }
+        }
+        return null
+    }
+
+    /**
+     * Build variant lookup table for efficient variant selection. Order of the variants matters,
+     * and this implementation mirrors the ordering provided by the local evaluation API.
+     */
+    private fun variantLookupTable(flag: FlagDefinition): List<VariantLookupEntry> {
+        val lookupTable = mutableListOf<VariantLookupEntry>()
+        var valueMin = 0.0
+
+        val variants = flag.filters.multivariate?.variants ?: emptyList()
```

Comment:
yo could use `?.let` here so you dont need to create a new instance of `emptyList()` if its null